src/test/java/com/bvan/vlad/MessageServiceTest.java

```diff
@@ -0,0 +1,149 @@
+        // then throw exception
+
+    }
+
+    @Test
+    public void lastMessageDate_whenListsMessagesInBatchesIsEmpty_thenCannotReturnInstant() {
+        // given
+        List<List<Message>> batches = new ArrayList<>();
+        List<Message> messages = new ArrayList<>();
+        batches.add(messages);
+        // when
+        assertThrows(NoSuchElementException.class, () -> messageService.lastMessageDate(batches));
+        // then throw exception
+
+    }
+
+    @Test
+    public void lastMessageDate_whenListsIsNull_thenCannotReturnInstant() {
+        // given
+        List<List<Message>> batches = new ArrayList<>();
+        batches.add(null);
+        // when
+        assertThrows(NoSuchElementException.class, () -> messageService.lastMessageDate(batches));
+        // then throw exception
+
+    }
+
+    @Test
+    public void lastMessageDate_whenOneBatchNullAndOtherNonNull_thenReturnInstant() {
+        // given
+        List<List<Message>> batches = new ArrayList<>();
+        for (int i = 1; i <= 11; i++) {
+            batches.add(getRandomBatch());
+        }
+        Message lastMessage = new Message("id", "last", Instant.now().plusSeconds(86400));
+        batches.get(0).add(lastMessage);
+        batches.add(null);
+        // when
+        Instant instant = messageService.lastMessageDate(batches);
+        // then
+        assertEquals(lastMessage.getCreatedDate(), instant);
+
+    }
+
+    @Test
+    public void lastMessageDate_whenOneMessageInListNullAndOtherNonNull_thenReturnInstant() {
+        // given
+        List<List<Message>> batches = new ArrayList<>();
+        for (int i = 1; i <= 11; i++) {
+            batches.add(getRandomBatch());
+        }
+        Message lastMessage = new Message("id", "last", Instant.now().plusSeconds(86400));
+        batches.get(0).add(lastMessage);
+        batches.get(1).add(null);
+        // when
+        Instant instant = messageService.lastMessageDate(batches);
+        // then
+        assertEquals(lastMessage.getCreatedDate(), instant);
+
+    }
+
+    @Test
+    public void lastMessageDate_whenInstantInOneMessageNull_thenReturnInstant() {
+        // given
+        List<List<Message>> batches = new ArrayList<>();
+        for (int i = 1; i <= 11; i++) {
+            batches.add(getRandomBatch());
+        }
+        Message lastMessage = new Message("id", "last", Instant.now().plusSeconds(86400));
+        batches.get(0).add(lastMessage);
+        batches.get(1).add(new Message("id", "body", null));
+        // when
+        Instant instant = messageService.lastMessageDate(batches);
+        // then
+        assertEquals(lastMessage.getCreatedDate(), instant);
+
+    }
+
+
+    private List<Message> getRandomBatch() {
+        List<Message> messages = new ArrayList<>();
+        for (int i = 1; i <= 10; i++) {
+
+            messages.add(new Message(Integer.toString(i), "body", Instant.now().plusSeconds(i * 10)));
+        }
+        return messages;
```

Comment:
Data generation adds additional complexity. In tests, simplicity is preferable.
```
List<List<Message>> batches = List.of(
  List.of(newMessage(2010), newMessage(2020)),
  List.of(newMessage(2015))
);
```
is preferable to
```
List<List<Message>> batches = generageData();
batches.get(0).add(new Message(...));
```